contentcuration/contentcuration/frontend/shared/vuex/file/actions.js

```diff
@@ -138,85 +121,127 @@ function hexToBase64(str) {
   );
 }
 
-export function uploadFileToStorage(context, { checksum, file, url, contentType }) {
-  return client.put(url, file, {
-    headers: {
-      'Content-Type': contentType,
-      'Content-MD5': hexToBase64(checksum),
-    },
-    onUploadProgress: progressEvent => {
-      context.commit('ADD_FILEUPLOAD', {
-        checksum,
-        loaded: progressEvent.loaded,
-        total: progressEvent.total,
+export function uploadFileToStorage(
+  context,
+  { id, file_format, mightSkip, checksum, file, url, contentType }
+) {
+  return (mightSkip ? client.head(storageUrl(checksum, file_format)) : Promise.reject())
+    .then(() => {
+      context.commit('ADD_FILE', {
+        id,
+        loaded: file.size,
+        total: file.size,
       });
-    },
-  });
+    })
+    .catch(() => {
+      return client
+        .put(url, file, {
+          headers: {
+            'Content-Type': contentType,
+            'Content-MD5': hexToBase64(checksum),
+          },
+          onUploadProgress: progressEvent => {
+            context.commit('ADD_FILE', {
+              id,
+              // Always assign loaded to a maximum of 1 less than the total
+              // to prevent progress being shown as 100% until the upload has
+              // completed
+              loaded: Math.min(progressEvent.loaded, progressEvent.total - 1),
+              total: progressEvent.total,
+            });
+          },
+        })
+        .then(() => {
+          // Set download progress to 100% now as we have confirmation of the
+          // completion of the file upload by the put request completing.
+          context.commit('ADD_FILE', {
```

Comment:
This would also be a good place to update the user's total used storage

Reply:
Hrm, except that (for the moment) this doesn't apply to Channel Thumbnails. Think we can add it as a follow up.